### PR TITLE
The Queue constructor should take an initial set of objects

### PR DESCRIPTION
**Pull request by chrisseaton**
https://bugs.ruby-lang.org/issues/17327

### CODE REVIEW

**Comment by chrisseaton** [comment]
Feedback from Matz was that this should accept an `Enumerable`, not `Array`.

**Comment by chrisseaton** [comment]
@tenderlove could you possibly review please? (I don't have rights to assign a reviewer.) Matz has approved in the attached ticket.

File: thread_sync.c [hunk 1]
```diff
@@ -836,18 +836,48 @@ queue_closed_result(VALUE self, struct rb_queue *q)
...
+    rb_ary_push(que, n);
+    return Qnil;
+}
+
 /*
  * Document-method: Queue::new
  *
- * Creates a new queue instance.
+ * Creates a new queue instance, optionally adding an initial contents of an
+ * Enumerable.
```
Comment (baweaver) [inline thread_sync.c:853]: Wording is a bit odd here, perhaps:

```
Creates a new queue instance, optionally using the contents of an Enumerable instance for its initial state
```

Comment (chrisseaton) [reply thread_sync.c:853]: Thanks, fixed.

**Comment by nobu** [comment]
`Set.new` seems unrelated.

**Comment by chrisseaton** [comment]
Are you asking why I've added specs for `Set#initialize`? I was asked to match the behaviour of `Set` rather than `Array`, but the behaviour of `Set` was under-specified, so I first expanded that and then copied for `Queue`.

File: spec/ruby/core/queue/initialize_spec.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+require_relative '../../spec_helper'
+
+describe "Queue#initialize" do
+  it "can be passed no arguments for an empty Queue" do
+    q = Queue.new
+    q.size.should == 0
+    q.empty?.should be_true
```
Comment (nobu) [inline spec/ruby/core/queue/initialize_spec.rb:7]:
```suggestion
    q.should.empty?
```

for better failure message, and also other `be_true`/`be_false` calls.

File: spec/ruby/core/queue/initialize_spec.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      enumerable.should_receive(:each_entry).and_yield(1).and_yield(2).and_yield(3)
+      q = Queue.new(enumerable)
+      q.size.should == 3
+      q.empty?.should be_false
+      q.pop.should == 1
+      q.pop.should == 2
+      q.pop.should == 3
+      q.empty?.should be_true
+    end
+
```
Comment (nobu) [inline spec/ruby/core/queue/initialize_spec.rb:32]:
```suggestion

```

File: spec/ruby/core/queue/initialize_spec.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    q = Queue.new
+    q.size.should == 0
+    q.empty?.should be_true
+  end
+
+  ruby_version_is '3.1' do
+    it "adds all elements of the passed Enumerable to self" do
+      q = Queue.new([1, 2, 3])
+      q.size.should == 3
+      q.empty?.should be_false
```
Comment (nobu) [inline spec/ruby/core/queue/initialize_spec.rb:14]:
```suggestion
      q.should_not.empty?
```

File: spec/ruby/core/queue/initialize_spec.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      q.pop.should == 1
+      q.pop.should == 2
+      q.pop.should == 3
+      q.empty?.should be_true
```
Comment (nobu) [inline spec/ruby/core/queue/initialize_spec.rb:18]:
```suggestion
      q.should.empty?
```

File: spec/ruby/core/queue/initialize_spec.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    end
+
+    it "uses #each_entry on the provided Enumerable" do
+      enumerable = MockObject.new('mock-enumerable')
+      enumerable.should_receive(:each_entry).and_yield(1).and_yield(2).and_yield(3)
+      q = Queue.new(enumerable)
+      q.size.should == 3
+      q.empty?.should be_false
```
Comment (nobu) [inline spec/ruby/core/queue/initialize_spec.rb:26]:
```suggestion
      q.should_not.empty?
```

File: spec/ruby/core/queue/initialize_spec.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      q.pop.should == 1
+      q.pop.should == 2
+      q.pop.should == 3
+      q.empty?.should be_true
```
Comment (nobu) [inline spec/ruby/core/queue/initialize_spec.rb:30]:
```suggestion
      q.should.empty?
```

File: spec/ruby/core/queue/initialize_spec.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    end
+  
+    it "uses #each on the provided Enumerable if it does not respond to #each_entry" do
+      enumerable = MockObject.new('mock-enumerable')
+      enumerable.should_receive(:each).and_yield(1).and_yield(2).and_yield(3)
+      q = Queue.new(enumerable)
+      q.size.should == 3
+      q.empty?.should be_false
```
Comment (nobu) [inline spec/ruby/core/queue/initialize_spec.rb:38]:
```suggestion
      q.should_not.empty?
```

File: spec/ruby/core/queue/initialize_spec.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+      q.pop.should == 1
+      q.pop.should == 2
+      q.pop.should == 3
+      q.empty?.should be_true
```
Comment (nobu) [inline spec/ruby/core/queue/initialize_spec.rb:42]:
```suggestion
      q.should.empty?
```

File: thread_sync.c [hunk 10]
```diff
@@ -836,18 +836,48 @@ queue_closed_result(VALUE self, struct rb_queue *q)
...
+ *    	q = Queue.new([a, b, c])
+ *    	q = Queue.new(items)
  */
 
 static VALUE
-rb_queue_initialize(VALUE self)
+rb_queue_initialize(int argc, VALUE *argv, VALUE self)
 {
+    VALUE initial;
+    ID id_enum_method;
```
Comment (nobu) [inline thread_sync.c:866]:
```suggestion
```

File: thread_sync.c [hunk 11]
```diff
@@ -836,18 +836,48 @@ queue_closed_result(VALUE self, struct rb_queue *q)
...
+    rb_scan_args(argc, argv, "01", &initial);
+    if (argc == 1) {
+        if (rb_respond_to(initial, id_each_entry)) {
+            id_enum_method = id_each_entry;
+        } else if (rb_respond_to(initial, id_each)) {
+            id_enum_method = id_each;
+        } else {
+            rb_raise(rb_eArgError, "value must be enumerable");
+        }
+        rb_block_call(initial, id_enum_method, 0, 0, push_to_queue, q->que);
```
Comment (nobu) [inline thread_sync.c:879]: `rb_check_block_call` checks if the method is defined then calls it at once.

```suggestion
        if (rb_check_block_call(initial, id_each_entry, 0, 0, push_to_queue, q->que) == Qundef &&
            rb_check_block_call(initial, id_each, 0, 0, push_to_queue, q->que) == Qundef) {
            rb_raise(rb_eArgError, "value must be enumerable");
        }
```

File: NEWS.md [hunk 12]
```diff
@@ -34,6 +34,11 @@ Outstanding ones only.
       You need to use a Hash literal to set a Hash to a first member.
       [[Feature #16806]]
 
+* Queue
+
+    * Queue#initialize now accepts an Enumerable of initial values.
+      [[Feature #17327]]
```
Comment (nobu) [inline NEWS.md:40]: You need to add `[Feature #17327]: https://bugs.ruby-lang.org/issues/17327` at the end of this file.

File: spec/ruby/library/set/initialize_spec.rb [hunk 13]
```diff
@@ -14,6 +14,31 @@
...
+    s.should include(1)
+    s.should include(2)
+    s.should include(3)
+  end
+
+  it "raises if the provided Enumerable does not respond to #each_entry or #each" do
+    enumerable = MockObject.new('mock-enumerable')
+    -> { Set.new(enumerable) }.should raise_error(ArgumentError, "value must be enumerable")
+  end
+
```
Comment (nobu) [inline spec/ruby/library/set/initialize_spec.rb:41]:
```suggestion
```

**Review by nobu — CHANGES_REQUESTED** [review]

File: spec/ruby/library/set/initialize_spec.rb (hunk 13 above)
Comment (eregon) [reply spec/ruby/library/set/initialize_spec.rb:41]: I think it's fine to add more specs, even if not directly testing the changed (yet it's related), but it should probably be its own commit, or its own PR.

File: thread_sync.c [hunk 14]
```diff
@@ -836,18 +836,48 @@ queue_closed_result(VALUE self, struct rb_queue *q)
  *
  */
 
+static ID id_each;
```
Comment (nobu) [inline thread_sync.c:839]: Predefined ID is available.

```suggestion
#define id_each idEach
```

File: thread_sync.c [hunk 15]
```diff
@@ -1613,6 +1643,8 @@ Init_thread_sync(void)
     rb_define_alloc_func(rb_cConditionVariable, condvar_alloc);
 
     id_sleep = rb_intern("sleep");
+    id_each = rb_intern("each");
```
Comment (nobu) [inline thread_sync.c:1646]:
```suggestion
```

**Review by nobu — CHANGES_REQUESTED** [review]

**Comment by nobu** [comment]
> Are you asking why I've added specs for `Set#initialize`? I was asked to match the behaviour of `Set` rather than `Array`, but the behaviour of `Set` was under-specified, so I first expanded that and then copied for `Queue`.

Yes, and I've thought that you would try `to_ary` first.
Anyway, I agree with @eregon's https://github.com/ruby/ruby/pull/3768#discussion_r566807617.

File: thread_sync.c [hunk 16]
```diff
@@ -836,18 +836,48 @@ queue_closed_result(VALUE self, struct rb_queue *q)
...
+rb_queue_initialize(int argc, VALUE *argv, VALUE self)
 {
+    VALUE initial;
+    ID id_enum_method;
     struct rb_queue *q = queue_ptr(self);
     RB_OBJ_WRITE(self, &q->que, ary_buf_new());
     list_head_init(queue_waitq(q));
+    rb_scan_args(argc, argv, "01", &initial);
+    if (argc == 1) {
+        if (rb_respond_to(initial, id_each_entry)) {
```
Comment (ko1) [inline thread_sync.c:872]: maybe calling `to_a` with `rb_convert_type` is enough.

**Comment by chrisseaton** [comment]
I've split the new specs for `Set#initialize` into another commit.

File: NEWS.md (hunk 12 above)
Comment (chrisseaton) [reply NEWS.md:40]: Thanks done.

File: spec/ruby/library/set/initialize_spec.rb (hunk 13 above)
Comment (chrisseaton) [reply spec/ruby/library/set/initialize_spec.rb:41]: Done in its own commit.

**Comment by nobu** [comment]
I think the commit for `Set` should be a separate issue, not in this one.

**Comment by chrisseaton** [comment]
Set specs moved to https://github.com/ruby/spec/pull/830.

**Pull request closed without merging**